File: dsa/java/src/sorting/MergeSort.java
package sorting;

import java.util.Arrays;

public class MergeSort {
    public static void main(String[] args){
        int[] list = new int[10];
        list[0] = 22;
        list[1] = 18;
        list[2] = 40;
        list[3] = 12;
        list[4] = 1;
        list[5] = 90;
        list[6] = 72;
        list[7] = 110;
        list[8] = 9;
        list[9] = 51;

        int[] newList = mergeSort(list);
        for(int item:newList){
            System.out.println(item);
        }
    }

    public static int[] mergeSort(int[] list){
        if(list.length == 1){
            return list;
        }

        int mid = (int) Math.floor(list.length / 2);
        int[] left = mergeSort(Arrays.copyOfRange(list, 0, mid));
        int[] right = mergeSort(Arrays.copyOfRange(list, mid, list.length));
        return merge(left, right);
    }

    public static int[] merge(int[] left, int[] right){
        
    }
}

File: dsa/mergeSort.ts
let list:number[] = [14, 33, 27, 10, 35, 19, 42, 44];
console.log(mergeSort(list));

function mergeSort(list:number[]){
    if(list.length <= 1) return list;

    let mid = Math.floor(list.length / 2);

    let list1 = mergeSort(list.slice(0, mid));
    let list2 = mergeSort(list.slice(mid));

    return merge(list1, list2);
}

function merge(list1:number[], list2:number[]){

    let temp:number[] = [];

    while(list1.length && list2.length){
        if(list1[0] > list2[0]){
            temp.push(list2.shift());
        }else{
            temp.push(list1.shift());
        }
    }
    console.log('temp is ' + temp);

    return temp.concat(list1.slice().concat(list2.slice()));
}